### PR TITLE
Preprocessor

### PR DESCRIPTION
**Pull request by sabel83**
This is a major new feature. If you see other ways to approach the problem (providing Preshell's functionality), don't hesitate to let me know.

### CODE REVIEW

File: docs/develop/rationales.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+## Preshell integration
+
+Preprocessor support has been added to Metashell to provide a
+[Preshell](https://github.com/sabel83/preshell) replacement. It has been added
+in a way that a set of seemingly meaningless switches have been added to the
+shell, which can be set in a way that turns Metashell into a Preprocessor REPL
+(Read-Eval-Print) shell. These settings are the following:
+
+* Echo the preprocessed version of each command (`#msh preprocessed echo on`)
```
Comment (r0mai) [inline docs/develop/rationales.md]: What about a command line flag which makes metashell start up in this mode?

File: test/system/include/metashell_system_test/pattern.hpp [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+#include <boost/variant.hpp>
+#include <boost/optional.hpp>
+
+#include <string>
+
+namespace metashell_system_test
+{
+  template <class T>
+  class pattern
```
Comment (r0mai) [inline test/system/include/metashell_system_test/pattern.hpp]: Can you explain what this class is for? I having a hard time figuring it out.

File: docs/develop/rationales.md (hunk 1 above)
Comment (r0mai) [inline docs/develop/rationales.md]: Sorry, wrong place for this comment

File: docs/manual/getting_started.md [hunk 3]
```diff
@@ -697,3 +697,85 @@ by calling `format` recursively on each element (see `format<Head>::type`).
...
 The version of the shell is also available for the headers. Metashell defines
 the `__METASHELL_MAJOR`, `__METASHELL_MINOR` and `__METASHELL_PATCH` macros.
+
+## Testing macros
+
+Metashell can be used as an interactive shell for testing macros. The easiest
+way to get started with it is to run the following command in Metashell:
+
+```cpp
+> #msh perprocessor mode
```
Comment (r0mai) [inline docs/manual/getting_started.md]: What about a command line flag which makes metashell start up in this mode?

**Comment by r0mai** [comment]
See comments, otherwise looks nice. Preshell benefits a lot from the misc improvements that was made to metashell but not ported to preshell. :+1:

File: test/system/include/metashell_system_test/pattern.hpp (hunk 2 above)
Comment (sabel83) [inline test/system/include/metashell_system_test/pattern.hpp]: I've added comments explaining it in 4f9639635af54179dc6a221ffb62a0232bf4173d. Let me know if this is still not clear.

Comment (r0mai) [inline test/system/include/metashell_system_test/pattern.hpp]: I see, it's clear now, thanks!

File: docs/manual/getting_started.md (hunk 3 above)
Comment (sabel83) [inline docs/manual/getting_started.md]: I've added `--preprocessor`.

**Pull request merged**